test/test_helper.rb

```diff
@@ -43,6 +45,26 @@ def enqueue_at(job, _delay)
 ActiveJob::Base.queue_adapter = :iteration_test
 
 class Product < ActiveRecord::Base
+  has_many :variants, dependent: :destroy_async
+end
+
+class SoftDeletedProduct < ActiveRecord::Base
+  self.table_name = "products"
+  has_many :variants, foreign_key: "product_id", dependent: :destroy_async, ensuring_owner_was: :deleted?
+
+  def deleted?
+    deleted
+  end
+
+  def destroy
+    update!(deleted: true)
+    run_callbacks(:destroy)
+    run_callbacks(:commit)
```

Comment:
It was not obvious for me why this was needed at first glance, figured I'd mention for reviewers: since https://github.com/rails/rails/pull/41093 the `before_destroy` callback adds the job class and params to the parent model `_after_commit_jobs` array, then the `after_commit` callback actually enqueues the job.